kcidb/monitor/spool/__init__.py

```diff
@@ -99,27 +99,43 @@ def post(self, notification, timestamp=None):
                             datetime.datetime.now(datetime.timezone.utc).
 
         Returns:
-            True, if the notification was posted onto the spool,
-            False, if not (it was already there).
+            True, if the new notification was posted onto the spool,
+            or the existing, "unpicked" notification was updated,
+            False, if not (it existed and was "picked" at the moment).
         """
         assert isinstance(notification, Notification)
         assert timestamp is None or \
-            isinstance(timestamp, datetime.datetime) and timestamp.tzinfo
+               isinstance(timestamp, datetime.datetime) and timestamp.tzinfo
         if timestamp is None:
             timestamp = datetime.datetime.now(datetime.timezone.utc)
 
-        try:
-            self._get_doc(notification.id).create(dict(
-                created_at=timestamp,
-                # Set to a definitely timed-out time, free for picking
-                picked_until=datetime.datetime.min,
-                message=notification.render().as_string(
-                    policy=email.policy.SMTPUTF8
-                ),
-            ))
-        except Conflict:
-            return False
-        return True
+        @firestore.transactional
+        def create_or_update(transaction):
+            message = notification.render(). \
+                as_string(policy=email.policy.SMTPUTF8)
+            due = notification.message.due \
+                if notification.message.due else timestamp
+            doc = self._get_doc(notification.id)
+            snapshot = doc.get(field_paths=["picked_until"],
+                               transaction=transaction)
+            if snapshot.exists:
+                # if the notification has been picked
+                if snapshot.get("picked_until") > timestamp:
+                    return False
+                transaction.update(doc, dict(message=message, due=due))
+            else:
+                transaction.create(doc, dict(
+                    created_at=timestamp,
+                    # Set to a definitely timed-out time, free for picking
+                    picked_until=datetime.datetime.min,
+                    message=notification.render().as_string(
+                        policy=email.policy.SMTPUTF8
+                    ),
```

Comment:
You're rendering the notification a second time here, and that's costly. Please change to render it once only.

Reply:
No I am not 
I have removed rendering notification from the post function, it is rendered only in the create_or_update clousre 

You requested that I moved it there yesterday 
From this comment: 

> Could you move this line into the create_or_update closure? Nothing outside it needs it, and this way we'll protect ourselves from accidentally adding some code accessing the doc outside that function, and thus outside the transaction, which is supposed to protect data consistency.

Reply:
Ah, that was about assigning the `doc` variable. I left a comment at the line where you're rendering the notification first. Here you're rendering it again, and it's better to avoid that, as that's a costly operation.

Reply:
> 



> 

Yes I have figured out the problem with due time 
This is the new update I have: https://github.com/kernelci/kcidb/pull/270/files#diff-cb11dfc89a15f32781b56c749d3b12fc5292e76f816d6575ef91a4c1ca6d52d7R116